src/Microsoft.DotNet.Build.Tasks/PackageFiles/ApiCompat.targets

```diff
@@ -1,137 +1,121 @@
-    <ReferenceAssembly>$(ReferenceOutputPath)\$(AssemblyName).dll</ReferenceAssembly>
-    <ReferenceProjectJson Condition="'$(ReferenceProjectJson)'=='' and Exists('$(ReferenceProjectDirectory)/project.json')">$(ReferenceProjectDirectory)/project.json</ReferenceProjectJson>
-    <ReferenceProjectLockJson Condition="Exists('$(ReferenceProjectJson)') and '$(ReferenceProjectLockJson)'==''">$(ReferenceProjectDirectory)/project.lock.json</ReferenceProjectLockJson>    
+    <ApiCompatBaseline Condition="!Exists('$(ApiCompatBaseline)')">$(MSBuildProjectDirectory)\ApiCompatBaseline.$(TargetGroup).txt</ApiCompatBaseline>
+    <ApiCompatBaseline Condition="!Exists('$(ApiCompatBaseline)')">$(MSBuildProjectDirectory)\ApiCompatBaseline.txt</ApiCompatBaseline>
   </PropertyGroup>
-  
+
   <PropertyGroup>
     <RunApiCompatForSrc Condition="$(MSBuildProjectDirectory.EndsWith('src'))">true</RunApiCompatForSrc>
     <RunApiCompatForRef Condition="$(MSBuildProjectDirectory.EndsWith('ref'))">true</RunApiCompatForRef>
     <RunApiCompat Condition="'$(RunApiCompat)'==''">false</RunApiCompat>
+
+    <ResolveMatchingContract Condition="'$(RunApiCompatForSrc)'=='true'">true</ResolveMatchingContract>
+    <TargetsTriggeredByCompilation Condition="'$(RunApiCompatForSrc)'=='true'">$(TargetsTriggeredByCompilation);ValidateApiCompatForSrc</TargetsTriggeredByCompilation>
+    <TargetsTriggeredByCompilation Condition="'$(RunApiCompatForRef)'=='true'">$(TargetsTriggeredByCompilation);ValidateApiCompatForRef</TargetsTriggeredByCompilation>
   </PropertyGroup>
-  
+
   <!-- ApiCompat for Implementation Assemblies  -->
   <Target Name="ValidateApiCompatForSrc"
-          AfterTargets="Build;CopyFilesToOutputDirectory"
-          DependsOnTargets="GetDependencyDirectories;GetContractDependencies" 
-          Condition="'$(RunApiCompatForSrc)' == 'true' AND '$(RunApiCompat)' == 'true'" >
-    <Warning Condition="!Exists('$(ReferenceProjectLockJson)') AND Exists('$(ReferenceProject)')" Text="The project.lock.json was not found for : $(ReferenceProject)" />    
-    <PropertyGroup Condition="Exists('$(ReferenceProjectLockJson)')">
+          DependsOnTargets="CopyFilesToOutputDirectory"
+          Condition="'$(RunApiCompatForSrc)' == 'true' AND '$(RunApiCompat)' == 'true' AND Exists('$(ContractProject)')" >
+
+    <PropertyGroup>
+      <ReferenceAssembly>@(ResolvedMatchingContract)</ReferenceAssembly>
+      <RefProjectForDependencies>$(ContractProject)</RefProjectForDependencies>
+      <!--
+        For depproj's we should approxiate the dependencies based on the current contract which
+        should be a superset of the older contract dependencies.
+      -->
+      <RefProjectForDependencies Condition="$(ContractProject.EndsWith('depproj'))">$(SourceDir)/$(AssemblyName)/ref/$(AssemblyName).csproj</RefProjectForDependencies>
+    </PropertyGroup>
+
+    <MSBuild Projects="$(RefProjectForDependencies)" Targets="ResolveAssemblyReferences">
```

Comment:
I added `OSGroup;TargetGroup` to the RemovedProperties to ensure those are at least consistent with the way we undefine them for contract ProjectReferences.